routing/dht/dht.go

```diff
@@ -71,15 +73,17 @@ func NewDHT(ctx context.Context, h host.Host, dstore ds.ThreadSafeDatastore) *Ip
 	// register for network notifs.
 	dht.host.Network().Notify((*netNotifiee)(dht))
 
-	dht.ContextGroup = ctxgroup.WithContextAndTeardown(ctx, func() error {
+	dht.proc = goprocessctx.WithContextAndTeardown(ctx, func() error {
 		// remove ourselves from network notifs.
 		dht.host.Network().StopNotify((*netNotifiee)(dht))
 		return nil
 	})
 
+	dht.ctx = ctx
+
 	h.SetStreamHandler(ProtocolDHT, dht.handleNewStream)
-	dht.providers = NewProviderManager(dht.Context(), dht.self)
-	dht.AddChild(dht.providers)
+	dht.providers = NewProviderManager(dht.ctx, dht.self)
+	dht.proc.AddChild(dht.providers.proc)
```

Comment:
not 100% this is safe.

```go
dht.proc = goprocessctx.WithContextAndTeardown(...)
// suppose ctx is cancelled here
dht.proc.AddChild(dht.providers.proc) // proc may be closed by here.
```
And will this panic. (we need way better semantics on the interaction with contexts)

i believe we want this:
```go
dht.proc = goprocess.WithTeardown(...) 
dht.proc.AddChild(dht.providers.proc)
// ok, now safe to close.
go func() {
  <-ctx.Done()
  dht.proc.Close()
}()
```

I just made `CloseAfterContext(p, ctx)` in https://github.com/jbenet/goprocess/pull/8 so just:

```go
dht.proc = goprocess.WithTeardown(...) 
dht.proc.AddChild(dht.providers.proc)
// ok, now safe to close.
goprocessctx.CloseAfterContext(dht.proc, ctx)
```